themes/default/content/blog/organizing-aws-accounts-with-pulumi/index.md

```diff
@@ -0,0 +1,290 @@
+   name: "developers",
+});
+
+new aws.iam.GroupPolicy("developersGroupPolicy", {
+   group: devGroup.name,
+   policy: {
+       Statement: [
+           {
+               Effect: "Allow",
+               Action: "sts:AssumeRole",
+               Resource: `arn:aws:iam::*:role/DeveloperRole`,
+           },
+       ],
+       Version: "2012-10-17",
+   },
+});
+```
+
+And create the IAM user that we’ll use for automation purposes. Again, you don’t need to do this if you plan to use OIDC federation for your CI/CD service.
+
+```typescript
+const automationUser = new aws.iam.User("automationUser", {
+   name: "cicd-automation",
+});
+
+new aws.iam.UserPolicy("automationUserPolicy", {
+   policy: {
+       Statement: [
+           {
+               Effect: "Allow",
+               Action: "sts:AssumeRole",
+               Resource: `arn:aws:iam::*:role/AutomationRole`,
+           },
+       ],
+       Version: "2012-10-17",
+   },
+   user: automationUser.name,
+});
+```
+
+## Member/Target Account Permissions
+
+We’ve created the base infrastructure in our root (aka management) account. We now need to create roles in the target account that allows users in the management account to assume.
+
+![Target Account Permissions](target-account-permissions.png)
+
+We’ll use a component resource to set up the necessary roles and policies in the target account. If you recall, when the target account was created, we specified a role name. That role name is the initial role that AWS creates on our behalf in the brand new account with admin privileges. The credentials we are using currently to execute our Pulumi app is assumed to be an admin user in the root account since we are dealing with highly-privileged operations here. We will need to create an explicit provider for our target account that can assume that automatically-created role.
+
+```typescript
+const devAccountProvider = new aws.Provider("devAccountProvider", {
+   allowedAccountIds: [devAccount.id],
+   assumeRole: {
+       roleArn: pulumi.interpolate`arn:aws:iam::${devAccount.id}:role/OrganizationalAccountAccessRole`,
+   },
+});
+```
+
+Now, whenever we want to modify resources in the target account in this app, we’ll use that provider. We’ll pass this provider to the component mentioned above, so that the roles and policies created by that component are performed using this provider instance. The component resource code is not included in this post for brevity. It provides a nice way to encapsulate the roles and policies we want to bootstrap our brand new account with. Of course, you can customize those roles and what permissions they get based on your use-case.
+
+```typescript
+const devAccountPermissions = new AccountPermissions(
+   "devAccountPermissions",
+   {
+       automationUser,
+       managementAccountId: pulumi.Output.create(
+           aws.getCallerIdentity().then((i) => i.accountId)
+       ),
+   },
+   { provider: devAccountProvider }
+);
+```
+
+## Backup Policies
```

Comment:
This is cool, but feels somewhat orthogonal to the content of the post, which is about account creation. Totally fine to leave it in -- I just figured I'd mention it as something to consider if you were thinking about making any revisions.

(Out of curiosity, though, if you do leave it in, it might be good to mention how it works, and sprinkle in a link or two to the AWS docs. For example, I sort of assumed I'd see something about what kinds of "things" would be backed up (DB, etc., as you mention), but that doesn't seem to be called out in the code.)

Reply:
If I'm understanding correctly, I _think_ the idea here is to also show the "other things" that need to be done when creating accounts within an AWS Organization. If that's the case, I'd suggest some sort of transition here ("In addition to creating and deleting accounts, most teams will also want to make sure that resources are backed up, tagging is enforced...").

Reply:
@cnunciato

> This is cool, but feels somewhat orthogonal to the content of the post, which is about account creation. Totally fine to leave it in -- I just figured I'd mention it as something to consider if you were thinking about making any revisions.

It's basically what @scottslowe mentions in his comment above. I think the suggestion to add a transition is a good one. I'll update this.

Reply:
I've updated this section and the next. Let me know what you think.